include/revng/Pipeline/Structure.h

```diff
@@ -0,0 +1,224 @@
+    const auto &Curr = *Steps.find(Current.getName());
+    if (not Curr.second.hasPredecessor())
+      return nullptr;
+    return &Steps.find(Curr.second.getPreviousStepName())->second.getStep();
+  }
+
+  const Step *getPredecessor(const Step &Current) const {
+    const auto &Curr = *Steps.find(Current.getName());
+    if (not Curr.second.hasPredecessor())
+      return nullptr;
+    return &Steps.find(Curr.second.getPreviousStepName())->second.getStep();
+  }
+
+  Step &getStep(llvm::StringRef Name) {
+    revng_assert(Steps.find(Name) != Steps.end());
+    return Steps.find(Name)->second.getStep();
+  }
+
+  const Step &getStep(llvm::StringRef Name) const {
+    revng_assert(Steps.find(Name) != Steps.end());
+    return Steps.find(Name)->second.getStep();
+  }
+
+  Step &operator[](llvm::StringRef Name) { return getStep(Name); }
+
+  const Step &operator[](llvm::StringRef Name) const { return getStep(Name); }
+
+  const ContainerSet &getContainers(llvm::StringRef Name) const {
+    return operator[](Name).getContainers();
+  }
+  ContainerSet &getContainers(llvm::StringRef Name) {
+    return operator[](Name).getContainers();
+  }
+
+  const ContainerSet &getStartingContainers() const {
+    return operator[](ReversePostOrderIndexes.front()).getContainers();
+  }
+  ContainerSet &getStartingContainers() {
+    return operator[](ReversePostOrderIndexes.front()).getContainers();
+  }
+
+  template<typename ContainerT>
+  const ContainerT &getStartingContainer(llvm::StringRef Name) const {
+    const auto &Containers = getStartingContainers();
+    return llvm::cast<ContainerT>(Containers.at(Name));
+  }
+
+  template<typename ContainerT>
+  ContainerT &getOrCreateStartingContainer(llvm::StringRef Name) {
+    auto &Containers = getStartingContainers();
+    return llvm::cast<ContainerT>(Containers[Name]);
+  }
+
+  template<typename ContainerT>
+  const ContainerT &
+  getContainer(llvm::StringRef Step, llvm::StringRef Name) const {
+    const auto &Containers = getContainers(Step);
+    return llvm::cast<ContainerT>(Containers.at(Name));
+  }
+
+  template<typename ContainerT>
+  ContainerT &getOrCreateContainer(llvm::StringRef Step, llvm::StringRef Name) {
+    auto &Containers = getContainers(Step);
+    return llvm::cast<ContainerT>(Containers[Name]);
+  }
+
+  llvm::Error
+  getInvalidations(llvm::StringMap<ContainersStatus> &Invalidated) const;
+
+  iterator begin() { return revng::mapToValueIterator(Steps.begin()); }
+  iterator end() { return revng::mapToValueIterator(Steps.end()); }
+
+  const_iterator begin() const {
+    return revng::mapToValueIterator(Steps.begin());
+  }
+  const_iterator end() const { return revng::mapToValueIterator(Steps.end()); }
+
+  llvm::Error getObjectives(llvm::StringRef EndingStepName,
+                            const ContainersStatus &Targets,
+                            ContainersStatus &ToLoad,
+                            std::vector<PipelineExecutionEntry> &ToExec);
```

Comment:
Is this method supposed to be public? What does it do? If it becomes private we can push `PipelineExecutionEntry` to the `.cpp` file.